tasks/inventory_targets.rb

```diff
@@ -0,0 +1,241 @@
+              "virtualmachines?api-version=2017-03-30"
+            else
+              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
+              "virtualmachines?api-version=2019-07-01"
+            end
+          else
+            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+            "providers/Microsoft.Compute/virtualmachines?api-version=2019-07-01"
+          end
+
+    vms = get_all_results(url, token)
+
+    # Filter by location
+    if opts[:location]
+      vms.select! { |vm| vm['location'] == opts[:location] }
+    end
+
+    # Filter by tags - tags are ANDed
+    if opts[:tags]
+      # Tag names are case insensitive, values are case sensitive
+      expected_tags = opts[:tags].map { |name, value| [name.to_s.downcase, value] }.to_h
+
+      vms.select! do |vm|
+        present_tags = vm.fetch('tags', {}).map { |name,value| [name.downcase, value] }.to_h
+        # Hash <= checks whether the lhs is a subset of the rhs
+        expected_tags <= present_tags
+      end
+    end
+
+    vms
+  end
+
+  def nics(token, creds, opts)
+    url = if opts[:resource_group]
+            if opts[:scale_set]
+              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
+              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
+              "networkinterfaces?api-version=2017-03-30"
+            else
+              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
+              "networkInterfaces?api-version=2019-07-01"
+            end
+          else
+            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+            "providers/Microsoft.Network/networkInterfaces?api-version=2019-07-01"
+          end
+
+    get_all_results(url, token)
+  end
+
+  def index_by_id(objects)
+    objects.each_with_object({}) do |obj, acc|
+      acc[obj['id']] = obj
+    end
+  end
+
+  # Uses the client credentials grant flow
+  # https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-client-creds-grant-flow
+  def token(creds)
+    data = {
+      grant_type: 'client_credentials',
+      client_id: creds['client_id'],
+      client_secret: creds['client_secret'],
+      resource: 'https://management.azure.com'
+    }
+
+    uri = URI.parse("https://login.microsoftonline.com/#{creds['tenant_id']}/oauth2/token")
+
+    request(:Post, uri, data)
+  end
+
+  def request(verb, uri, data, header = {})
+    # Create the client
+    client = Net::HTTP.new(uri.host, uri.port)
+
+    # Azure REST API always uses SSL
+    client.use_ssl = true
+    client.verify_mode = OpenSSL::SSL::VERIFY_PEER
```

Comment:
Without a `require 'openssl'` the task errors for me. 
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt command run whoami -t all --debug
[snip]
"kind":"NameError","msg":"uninitialized constant AzureInventory::OpenSSL","details":["/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/azure_inventory/tasks/inventory_targets.rb:196:in `request'","/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/azure_inventory/tasks/inventory_targets.rb:187:in `token'","/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/azure_inventory/tasks/inventory_targets.rb:13:in `inventory_targets'","/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/azure_inventory/tasks/inventory_targets.rb:231:in `task'","/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/ruby_task_helper/files/task_helper.rb:49:in `run'","/tmp/b3e56880-2c43-4375-8fbf-a35f72067102/azure_inventory/tasks/inventory_targets.rb:241:in `<main>'"],"_error":{"kind":"puppetlabs.tasks/task-error","issue_code":"TASK_ERROR","msg":"The task failed with exit code 1","details":{"exit_code":1}}}}
Error executing plugin: task from inventory_targets in azure-vms: The task failed with exit code 1
```

Reply:
I guess I only ever ran it on localhost, where maybe the behavior is different. Fixed.